### PR TITLE
apteryx_timestamp: support refreshers / providers

### PR DESCRIPTION
**Pull request by blairsteven**
apteryx_timestamp returns a value that can be used to detect changes in
the database. When we have providers, a read of that value will always
be required, with refreshers this is required some period after the
previous update.

apteryx_timestamp will now return the current time when a provided value
timestamp is requested, and when an expired refreshed value is
requested. This value will be less than the subsequent data read, as is
required by the interface.

### CODE REVIEW

File: apteryxd.c [hunk 1]
```diff
@@ -1518,7 +1527,7 @@ collect_provided_paths(const char *_path, GNode *root)
     gchar *path = _path ? g_strdup(_path) : apteryx_node_path(root);
     GList *result = NULL;
     GList *provided_paths = NULL;
-
+    gchar *needle = g_strdup_printf("%s/", path);
```
Comment (carlgsmith) [inline apteryxd.c:1530]: why this change - doesn't it create a memory leak if !config_tree_has_providers

File: apteryxd.c [hunk 2]
```diff
@@ -1956,7 +1965,15 @@ handle_timestamp (rpc_message msg)
...
         /* Lookup value */
-        value = db_timestamp (path);
+        if (config_tree_has_providers (path) ||
+            call_refreshers (path, true))
+        {
+            value = calculate_timestamp ();
+        }
+        else
+        {
+            value = db_timestamp (path);
```
Comment (carlgsmith) [inline apteryxd.c:1975]: Current behaviour is that if anything is stored in the DB, then that trumps anything provided. I wonder if this should be
```
if call_refreshers
  calculate_timestamp
else if value in db
  db_timestamp
else if has_providers
  calculate_timestamp
```

**Comment by blairsteven** [comment]
That's a stray change from my experiments sorry. Needs removing.
